### PR TITLE
MDS208 | Implement authentication model in conjunction with scope handler pattern

### PR DESCRIPTION
**Pull request by ajchilts**
MDS208 | Implement authentication model in conjunction with scope handler pattern

### CODE REVIEW

File: app/uk/gov/hmrc/individualsdetailsapi/play/RequestHeaderUtils.scala [hunk 1]
```diff
@@ -16,6 +16,7 @@
 
 package uk.gov.hmrc.individualsdetailsapi.play
 
+import play.api.Logger
```
Comment (grab-a-byte) [inline app/uk/gov/hmrc/individualsdetailsapi/play/RequestHeaderUtils.scala]: Is this logger used?

File: app/uk/gov/hmrc/individualsdetailsapi/services/PathTree.scala [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+  private[service] def +(node: Node): B = B(value, Seq(node))
+
+  private[service] def ++(nodes: Iterable[Node]): B =
+    this + nodes.head ++ nodes.tail
+
+  override def toString: String = value
+}
+
+case class B(value: String, children: Iterable[Node]) extends Node {
```
Comment (grab-a-byte) [inline app/uk/gov/hmrc/individualsdetailsapi/services/PathTree.scala:80]: Could we have more meaningful names other than L and B?

**Review by grab-a-byte — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/individualsdetailsapi/services/PathTree.scala (hunk 2 above)
Comment (ajchilts) [reply app/uk/gov/hmrc/individualsdetailsapi/services/PathTree.scala:80]: @parkeradam - this is a straight copy of @owen-opencast handler pattern from the other service. 

Could likely come up with something but would need to be uniform across all services.

File: app/uk/gov/hmrc/individualsdetailsapi/play/RequestHeaderUtils.scala (hunk 1 above)
Comment (ajchilts) [reply app/uk/gov/hmrc/individualsdetailsapi/play/RequestHeaderUtils.scala]: Can be removed if not 👍

File: app/uk/gov/hmrc/individualsdetailsapi/services/PathTree.scala (hunk 2 above)
Comment (grab-a-byte) [reply app/uk/gov/hmrc/individualsdetailsapi/services/PathTree.scala:80]: Ok, I'll resolve for now but think we may need to bash heads and come up with some better names at some point.

File: app/uk/gov/hmrc/individualsdetailsapi/play/RequestHeaderUtils.scala (hunk 1 above)
Comment (grab-a-byte) [reply app/uk/gov/hmrc/individualsdetailsapi/play/RequestHeaderUtils.scala]: I just made the change to resolve.

**Review by grab-a-byte — APPROVED** [review]
LGTM

File: app/uk/gov/hmrc/individualsdetailsapi/services/ScopesService.scala [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+      .toMap
+
+  def getEndPointScopes(endpointKey: String): Iterable[String] = {
+    val keys = apiConfig
+      .getEndpoint(endpointKey)
+      .map(endpoint => endpoint.fields.keys.toList.sorted)
+      .getOrElse(List())
+
+    apiConfig.scopes
+      .filter(s => s.fields.toSet.subsetOf(keys.toSet))
```
Comment (owen-gilfellon) [inline app/uk/gov/hmrc/individualsdetailsapi/services/ScopesService.scala]: This doesn't work if a scope has fields from multiple endpoints (e.g. a default scope) - definition of this.subsetOf(that) is "true if this set is a subset of that, i.e. if every element of this set is also an element of that".... suggest something like:
`.filter(s => s.fields.toSet.intersect(keys.toSet).nonEmpty)`

File: app/uk/gov/hmrc/individualsdetailsapi/domains/ErrorResponses.scala [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    extends ErrorResponse(NOT_FOUND,
+                          "NOT_FOUND",
+                          "The resource can not be found")
+case object ErrorTooManyRequests
+    extends ErrorResponse(TOO_MANY_REQUESTS,
+                          "TOO_MANY_REQUESTS",
+                          "Rate limit exceeded")
+
+class MatchNotFoundException extends RuntimeException
+class ValidationException(message: String) extends RuntimeException(message)
```
Comment (owen-gilfellon) [inline app/uk/gov/hmrc/individualsdetailsapi/domains/ErrorResponses.scala:49]: ScopeAuthorisationException and ValidationException not used

**Review by owen-gilfellon — CHANGES_REQUESTED** [review]
Just a couple of small changes. Test coverage is down slightly (mostly the exceptions in CommonController that aren't being triggered), think that can be fixed later though.

**Review by owen-gilfellon — APPROVED** [review]

**Pull request merged**